src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,242 @@
+        public Task OnProjectLoaded()
+        {
+            return InitializeAsync();
+        }
+
+        public async Task<string[]> GetDesignTimeOutputFilenamesAsync()
+        {
+            var property = await _unconfiguredProjectServices.ActiveConfiguredProjectProperties.GetConfiguredBrowseObjectPropertiesAsync();
+            var basePath = await property.FullPath.GetValueAsPathAsync(false, false);
+            return AppliedValue.Value.Inputs.Keys.Select(i => Path.Combine(basePath, i)).ToArray();
+        }
+
+        public async Task<string> GetTempPEBlobAsync(string fileName)
+        {
+            var inputs = AppliedValue.Value;
+            if (fileName == null || inputs.Inputs.TryGetValue(fileName, out var cancellation))
+            {
+                return null;
+            }
+
+            // getting the next token will automatically cancel any current compilation that is already happening for this file
+            var token = cancellation.CreateNext();
+
+            await _unconfiguredProjectServices.ThreadingService.SwitchToUIThread(token);
+
+            token.ThrowIfCancellationRequested();
+
+            var property = await _unconfiguredProjectServices.ActiveConfiguredProjectProperties.GetConfiguredBrowseObjectPropertiesAsync();
+            var basePath = await property.FullPath.GetValueAsPathAsync(false, false);
+            var objPath = await property.IntermediatePath.GetValueAsPathAsync(false, false);
+            var outputPath = Path.Combine(basePath, objPath, "TempPE");
+
+            var files = new HashSet<string>(inputs.SharedInputs.Count + 1, StringComparers.Paths);
+            files.AddRange(inputs.SharedInputs.Select(i => Path.Combine(basePath, i)));
+            files.Add(Path.Combine(basePath, fileName));
+
+            var outputFileName = fileName + ".dll";
+
+            // var result = await _compiler.CompileAsync(_languageServiceHost.ActiveProjectContext, Path.Combine(outputPath, outputFileName), files, token);
+            //
+            // if (!result)
+            // {
+            //     return null;
+            // }
+
+            // VSTypeResolutionService is the only consumer, and it only uses the codebase element so just default most of them
+            return $@"<root>
+  <Application private_binpath = ""{outputPath}""/>
+  <Assembly
+    codebase = ""{outputFileName}""
+    name = ""{fileName}""
+    version = ""0.0.0.0""
+    snapshot_id = ""1""
+    replaceable = ""True""
+  />
+</root>";
+        }
+
+        protected override Task ApplyAsync(DesignTimeInputsDelta value)
+        {
+            var applied = AppliedValue.Value;
+            var designTimeInputs = applied.Inputs.ToBuilder();
+            var designTimeSharedInputs = applied.SharedInputs.ToBuilder();
+
+            foreach (var item in value.AddedItems)
+            {
+                // TODO: Raise events
+
+                designTimeInputs.Add(item, new CancellationSeries());
```

Comment:
Another TODO here would be to move this `new CancellationSeries` to `GetTempPEBlob` so that its lazy, and only created if we need to ever compile a TempPE DLL, but I want to get the tests written for that first so I can make sure that calling ApplyAsync from that method works as I expect it to.